File: 2-AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    curr1 = l1
    curr2 = l2
    carry = 0
    tot = 0
    dummyHead = ListNode()
    newCurr = dummyHead
    while curr1 or curr2:
      tot += carry
      if curr1:
        tot += curr1.val
        curr1 = curr1.next
      if curr2:
        tot += curr2.val
        curr2 = curr2.next
      carry = tot // 10
      tot = tot % 10

      newCurr.next = ListNode(tot)
      newCurr = newCurr.next
      tot = 0
    
    if carry > 0:
      newCurr.next = ListNode(carry)
      newCurr = newCurr.next

    return dummyHead.next

File: 2-AddTwoNumbers.java


class Solution {
  //Definition for singly-linked list.
  public class ListNode {
      int val;
      ListNode next;
      ListNode() {}
      ListNode(int val) { this.val = val; }
      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
  }
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode a = l1;
        ListNode b = l2;    
        int sum = (a.val + b.val);
        int carry = sum / 10;
        ListNode sumStart = new ListNode(sum % 10);    
        a = a.next;
        b = b.next;
        ListNode sumNode = sumStart;
        int intA;
        int intB;
        
        
        while (carry != 0 || a != null || b != null) {
            intA = 0;
            intB = 0;
            if (a != null) {
                intA = a.val;
                a = a.next;
            }
            if (b != null) {
                intB = b.val;
                b = b.next;
            }
            
            sum = (intA + intB + carry);
            sumNode.next = new ListNode(sum % 10);
            carry = sum / 10;
            sumNode = sumNode.next;
        }
        return sumStart;
    }
}